lib/util.js

```diff
@@ -1057,3 +1057,75 @@ exports._exceptionWithHostPort = function(err,
 // process.versions needs a custom function as some values are lazy-evaluated.
 process.versions[exports.inspect.custom] =
   (depth) => exports.format(JSON.parse(JSON.stringify(process.versions)));
+
+
+/**
+ * @typedef {Object} callbackifyOptions
+ * @property {function} transformer
+ * @property {boolean} toSplat
+ * @property {Array.<String>} project
+ */
+
+/**
+ * @param {function(): Promise} orig
+ * @param {callbackifyOptions} options
+ */
+function callbackify(orig, options) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to callbackify() ' +
+                        'must be a function');
+  }
+
+  // we have two way to convert the returned value into multiple arguments
+  // either splat it or project some of it's keys
+  if ('toSplat' in options && 'project' in options)
+    throw new TypeError('callbackify options can specify either ' +
+                        '`toSplat` or `project`, but not both');
+  const toSplat = options.toSplat;
+  const project = options.project;
+
+  const fn = function(...args) {
+    const cb = args.pop();
+    if (typeof cb !== 'function') {
+      throw new TypeError('The first argument to the callbackified function ' +
+        'must be a callback style function');
+    }
+    var p = orig.call(this, ...args);
+    if (typeof options.transformer === 'function') {
+      p = options.transformer(p);
+    }
+    p = p.then((ret) => {
+      if (toSplat) {
+        cb(null, ...ret);
+      } else if (project) {
+        const allThere = project.all((k) => k in ret);
+        if (!allThere) {
+          const notAll = new TypeError('Not all of the projected properties' +
+                                       ' are present in the returned value:' +
+                                       inspect(ret));
+          return cb(notAll);
+        } else {
+          const args = project.reduce((s, k) => s.concat(ret[k]), []);
+          cb(null, ...args);
+        }
+      } else {
+        cb(null, ret);
+      }
+    }, (rej) => {
+      // !rej guard inspired by BlueBird https://goo.gl/t5IS6M
+      if (!rej) {
+        const newRej = new Error(rej + '');
+        newRej.cause = rej;
+        rej = newRej;
+      }
+      cb(rej);
+    }).catch((e) => process.emit('uncaughtException', e));
+    return p;
```

Comment:
Basically, people were doing things like the following with BB:

```js
await promise.asCallback((err, data) => {
  console.log("2");
});
console.log("1");
```
And expected 2, 1 to log (and got 1,2), not to mention people thinking that if they nest callbacks it will somehow magically wait for them.